File: partition.cpp
#include <cstdio>
#include <cstdlib>
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <fstream>
#include <vector>
#include <time.h>
#include <math.h>
#include <cmath>
#include <ctime>
#include "maxheap.hpp"

#include <random>
#include <chrono>

using namespace std;

// HELPER METHODS

// Karmarkar Karp's algorithm
long karmarkarKarp(std::vector<long> testvec) {

    MaxHeap h;
    for (int i = 0; i < testvec.size(); i++) {
        h.insert(testvec[i]);
    }

    while (h.size() > 1) {
        long a = h.extractMax();
        long b = h.extractMax();
        h.insert(abs(a-b));
    }
    return h.extractMax();
}

// generates a random sequence of 1, -1 as a random solution
std::vector<int> generateRandSol(int n) {

    std::vector<int> sol(n);

    for (int i = 0; i < n; i++) {
        int set = rand() & 1;
		sol[i] = (set != 1) ? -1 : 1;
    }

    return sol;

}

// calculates residue
long residue(std::vector<long> nums, std::vector<int> sol, int n) 
{
    long res = 0;
    for (int i = 0; i < n; i++) {
        res += sol[i] * nums[i];
    }
    return abs(res);
}

// generates neighbor solution
std::vector<int> generateNeighbor(std::vector<int> sol) {

    int n = sol.size();

    std::vector<int> neighbor(n);

    int r = rand() % n;

	for (int i = 0; i < n; i++) {
		neighbor[i] = sol[i];
	}

	
	neighbor[r] = -neighbor[r];

	int j = r;
	if ((double) rand() / RAND_MAX <= 0.5) {
		while (j == r) 
        {
			j = rand() % n;
		}
		neighbor[j] = -neighbor[j];
	}

	return neighbor;
}

// for simulated annealing
double cooling(int ITER) 
{
    return pow(10, 10) * pow(0.8, ITER / 300);
}

// generates a random prepartitioning )
std::vector<int> generatePartition(int n) {
    
    std::vector<int> partition(n);
    for (int i = 0; i < n; i++) 
    {
        partition[i] = rand() % n;
    }

    return partition;
}

// calculates residue but for prepartitioning
long residue_pre(std::vector<long> nums, std::vector<int> sol, int n) {
    std::vector<long> modified_sol(n);

    for (int i = 0; i < n; i++) {
        modified_sol[i] = 0;
    }

    for (int i = 0; i < n; i++) {
        modified_sol[sol[i]] += nums[i];
    }

    return karmarkarKarp(modified_sol);
}

std::vector<int> neighbor_pre(std::vector<int> sol) {

    int n = sol.size();

    std::vector<int> neighbor(n);

	for (int i = 0; i < n; i++) {
		neighbor[i] = sol[i];
	}

    int r = rand() % n;

	int j = neighbor[r];
	while (j == neighbor[r]) {
		j = rand() % n;
	}
	neighbor[r] = j;

	return neighbor;

}


// HEURISTICS


// repeated random heuristic
long repeatedRandom(std::vector<long> testvec) {
    
    int n = testvec.size();

    std::vector<int> sol = generateRandSol(n);

    for (int i = 0; i < 25000; i++) {
        std::vector<int> sol2 = generateRandSol(n);
        if (residue(testvec, sol2, n) < residue(testvec, sol, n)) {
            sol = sol2;
        }
    }

    return residue(testvec, sol, n);

}

// hill climbing heuristic
long hillClimbing(std::vector<long> testvec) {

    int n = testvec.size();

    std::vector<int> sol = generateRandSol(n);

    for (int i = 0; i < 25000; i++) {
        std::vector<int> sol2 = generateNeighbor(sol);
        if (residue(testvec, sol2, n) < residue(testvec, sol, n)) {
            sol = sol2;
        }
    }

    return residue(testvec, sol, n);

}

// simulated annealing heuristic
long simulatedAnnealing(std::vector<long> testvec) {

    int n = testvec.size();

    std::vector<int> sol = generateRandSol(n);
    std::vector<int> sol3 = sol;

    for (int i = 0; i < 25000; i++) {
        std::vector<int> sol2 = generateNeighbor(sol);

        long res2 = residue(testvec, sol2, n);
        long res1 = residue(testvec, sol, n);
        if (res2 < res1) {
            sol = sol2;
        } else if ((double) rand() / RAND_MAX < exp(- (long) (res2 - res1) / cooling(i))) {
            sol = sol2;
        }

        if (res1 < residue(testvec, sol3, n)) {
            sol3 = sol;
        }
    }

    return residue(testvec, sol3, n);

}

std::vector<long> generateTest(){

    std::vector<long> testvec;
    srand(time(0));

    for (int i = 0; i < 100; i++) {

        long range = 1000000000000LL; // set the range [1, 10^12]
        long randNum = rand() % range + 1;

        testvec.push_back(randNum);
    }

    return testvec;

}

// repeated random heuristic w/ pre-partitioning
long rr_pre(std::vector<long> testvec) {
    
    int n = testvec.size();

    std::vector<int> sol = generatePartition(n);

    for (int i = 0; i < 25000; i++) {
        std::vector<int> sol2 = generateRandSol(n);
        if (residue_pre(testvec, sol2, n) < residue_pre(testvec, sol, n)) {
            sol = sol2;
        }
    }

    return residue_pre(testvec, sol, n);

}

void test() {

    long kk = 0;
    long rr = 0;
    long hc = 0;
    long sa = 0;
    long rrp = 0;

    for (int i = 0; i < 10; i++) {
        std::vector<long> testvec = generateTest();
        kk += karmarkarKarp(testvec);
        rr += repeatedRandom(testvec);
        hc += hillClimbing(testvec);
        sa += simulatedAnnealing(testvec);
        rrp += rr_pre(testvec);
    }

    std::cout << "Karmarkar Karp: " << kk / 10 << std::endl;
    std::cout << "Repeated Random: " << rr / 10 << std::endl;
    std::cout << "Hill Climbing: " << hc / 10 << std::endl;
    std::cout << "Simulated Annealing: " << sa / 10 << std::endl;
    std::cout << "Repeated Random with Pre-Partitioning: " << rrp / 10 << std::endl;


}


int main(int argc, char* argv[]) {

    // test 

    // command line argument should be in form "./partition flag algorithm inputfile"
    /*if (argc != 4) {
        printf("Usage: ./partition flag algorithm inputfile");
        return 0;
    }

    std::vector<long> testvec;
    if (atoi(argv[1]) == 0) {
        ifstream testfile(argv[3]);
        string line;
        while (getline(testfile, line)) {   
            testvec.push_back(stoi(line));
        }
    }


    if (atoi(argv[2]) == 0) {
        MaxHeap test;
        for (int i = 0; i < testvec.size(); i++) {
            test.insert(testvec[i]);
        }
        long n = karmarkarKarp(test);
        std::cout << n << std::endl;
    } else if (atoi(argv[2]) == 1) {
        long n = repeatedRandom(testvec);
        std::cout << n << std::endl;
    } else if (atoi(argv[2]) == 2) {
        long n = hillClimbing(testvec);
        std::cout << n << std::endl;
    } else if (atoi(argv[2]) == 3) {
        long n = simulatedAnnealing(testvec);
        std::cout << n << std::endl;
    }*/


    test();

    

}

File: partition.py
import sys
import random
from random import randint
import math
import heapq
import numpy as np
import matplotlib.pyplot as plt
import time

MAX_ITER = 25000

# HELPER FUNCTIONS

def generate_rand_sol(n):
    sol = []
    for i in range(n):
        random_choice = random.choice([-1,1])
        sol.append(random_choice)
    return sol

def calc_residues(nums, sol, n):
    sum = 0
    for i in range(n):
        sum += sol[i] * nums[i]
    return abs(sum)

def generate_neighbor(sol):
    n = len(sol)
    neighbor = sol.copy()
    rand_indices = random.sample(range(n),2)
    i = rand_indices[0]
    j = rand_indices[1]
    neighbor[i] = -neighbor[i]
    if random.random() < 0.5:
        neighbor[j] = -neighbor[j]
    return neighbor

def KK(nums):
    H = [-x for x in nums]
    heapq.heapify(H)
    while(len(H) > 1):
        x = heapq.heappop(H)
        y = heapq.heappop(H)
        heapq.heappush(H, x - y)
    return -heapq.heappop(H)

def RR(nums):
    n = len(nums)
    sol = generate_rand_sol(n)
    for _ in range(MAX_ITER):
        sol2 = generate_rand_sol(n)
        if calc_residues(nums, sol2, n) < calc_residues(nums, sol, n):
            sol = sol2.copy()
    #print("RR result:", calc_residues(nums, sol, n))
    return calc_residues(nums, sol, n)

def HC(nums):
    n = len(nums)
    sol = generate_rand_sol(n)
    for _ in range(MAX_ITER):
        sol2 = generate_neighbor(sol)
        if calc_residues(nums, sol2, n) < calc_residues(nums, sol, n):
            sol = sol2.copy()
    #print("HC result:", calc_residues(nums, sol, n))
    return calc_residues(nums, sol, n)

def T(iter):
    return pow(10, 10) * pow(.8, iter // 300)

def SA(nums):
    n = len(nums)
    sol = generate_rand_sol(n)
    sol3 = sol.copy()
    for i in range(MAX_ITER):
        sol2 = generate_neighbor(sol)
        if calc_residues(nums, sol2, n) < calc_residues(nums, sol, n):
            sol = sol2.copy()
        elif random.random() < pow(math.e, -(calc_residues(nums, sol2, n) - calc_residues(nums, sol, n))/T(i)):
            sol = sol2.copy()
        if calc_residues(nums, sol, n) < calc_residues(nums, sol3, n):
            sol3 = sol.copy()
    #print("SA result:", calc_residues(nums, sol3, n))
    return calc_residues(nums, sol3, n)

# PRE-PARTITIONING STUFF

def generate_partition(n):
    partition = []
    for i in range(n):
        rand_int = random.randint(0, n - 1)
        partition.append(rand_int)
    return partition

def p_generate_neighbor(sol):
    n = len(sol)
    neighbor = sol.copy()
    rand_indices = random.sample(range(n), 2)
    i = rand_indices[0]
    j = rand_indices[1]
    neighbor[i] = j
    return neighbor

def p_calc_residues(nums, partition, n):
    A_prime = [0] * (n)
    for i in range(n):
        A_prime[partition[i]] += nums[i]
    return KK(A_prime)

def p_RR(nums):
    n = len(nums)
    sol = generate_partition(n)
    for _ in range(MAX_ITER):
        sol2 = generate_partition(n)
        if p_calc_residues(nums, sol2, n) < p_calc_residues(nums, sol, n):
            sol = sol2.copy()
    return p_calc_residues(nums, sol, n)

def p_HC(nums):
    n = len(nums)
    sol = generate_partition(n)
    for _ in range(MAX_ITER):
        sol2 = p_generate_neighbor(sol)
        if p_calc_residues(nums, sol2, n) < p_calc_residues(nums, sol, n):
            sol = sol2.copy()
    return p_calc_residues(nums, sol, n)

def p_SA(nums):
    n = len(nums)
    sol = generate_partition(n)
    sol3 = sol.copy()
    for i in range(MAX_ITER):
        sol2 = p_generate_neighbor(sol)
        if p_calc_residues(nums, sol2, n) < p_calc_residues(nums, sol, n):
            sol = sol2.copy()
        elif random.random() < pow(math.e, -(p_calc_residues(nums, sol2, n) - p_calc_residues(nums, sol, n))/T(i)):
            sol = sol2.copy()
        if p_calc_residues(nums, sol, n) < p_calc_residues(nums, sol3, n):
            sol3 = sol.copy()
    return p_calc_residues(nums, sol3, n)



def main(args):

    if len(args) != 3:
        print("Not enough arguments.")
        return
    
    nums = []
    try:
        file = open(args[2], 'r')
        for num in file:
            nums.append(int(num))
        file.close()
    except:
        print("Error when reading input.")

    algorithm_type = int(args[1])
    match algorithm_type:
        case 0:
            print(KK(nums))
        case 1:
            print(RR(nums))
        case 2:
            print(HC(nums))
        case 3:
            print(SA(nums))
        case 11:
            print(p_RR(nums))
        case 12:
            print(p_HC(nums))
        case 13:
            print(p_SA(nums))
        


if __name__ == "__main__":
    main(sys.argv[1:])

    """
    kk_avg = 0
    kk_data = np.array([0]*50)
    for i in range(50):
        nums = np.array([randint(0, 10 ** 12) for _ in range(100)])
        start = time.perf_counter()
        kkval = KK(nums)
        end = time.perf_counter()
        elapsed = end - start
        kk_avg += elapsed
        kk_data[i] = elapsed

    print("KK time: ")
    print(kk_avg / 50)
    """
    """
    rr_avg = 0
    rr_data = np.array([0]*50)
    for i in range(50):
        nums = np.array([randint(0, 10 ** 12) for _ in range(100)])
        start = time.perf_counter()
        rrval = RR(nums)
        end = time.perf_counter()
        elapsed = end - start
        rr_avg += elapsed
        rr_data[i] = elapsed
    print("RR time: ")
    print(rr_avg / 50)

    hc_avg = 0
    hc_data = np.array([0]*50)
    for i in range(50):
        nums = np.array([randint(0, 10 ** 12) for _ in range(100)])
        start = time.perf_counter()
        hcval = HC(nums)
        end = time.perf_counter()
        elapsed = end - start
        hc_avg += elapsed
        hc_data[i] = elapsed
    print("HC time: ")
    print(hc_avg / 50)
    
    sa_avg = 0
    sa_data = np.array([0]*50)
    for i in range(50):
        nums = np.array([randint(0, 10 ** 12) for _ in range(100)])
        start = time.perf_counter()
        saval = SA(nums)
        end = time.perf_counter()
        elapsed = end - start
        sa_avg += elapsed
        sa_data[i] = elapsed
    print("SA time: ")
    print(sa_avg / 50)
    """
    """
    p_rr_avg = 0
    p_rr_data = np.array([0]*50)
    for i in range(50):
        nums = np.array([randint(0, 10 ** 12) for _ in range(100)])
        start = time.perf_counter()
        rrval = p_RR(nums)
        end = time.perf_counter()
        elapsed = end - start
        p_rr_avg += elapsed
        p_rr_data[i] = elapsed
    print("p_RR time: ")
    print(p_rr_avg / 50)

    p_hc_avg = 0
    p_hc_data = np.array([0]*50)
    for i in range(50):
        nums = np.array([randint(0, 10 ** 12) for _ in range(100)])
        start = time.perf_counter()
        hcval = p_HC(nums)
        end = time.perf_counter()
        elapsed = end - start
        p_hc_avg += elapsed
        p_hc_data[i] = elapsed
    print("p_HC time: ")
    print(p_hc_avg / 50)
    
    p_sa_avg = 0
    p_sa_data = np.array([0]*50)
    for i in range(50):
        nums = np.array([randint(0, 10 ** 12) for _ in range(100)])
        start = time.perf_counter()
        saval = p_SA(nums)
        end = time.perf_counter()
        elapsed = end - start
        p_sa_avg += elapsed
        p_sa_data[i] = elapsed
    print("p_SA time: ")
    print(p_sa_avg / 50)
    

    y5 = p_rr_data
    y6 = p_hc_data
    y7 = p_sa_data


    x = np.arange(1, 51)
    y1 = kk_data
    #y2 = rr_data
    #y3 = hc_data
    #y4 = sa_data

    plt.title("Comparison of Times for KK and Other Heuristics (in Sec)")
    plt.xlabel("Instance")
    plt.ylabel("Time Elapsed (seconds)")
    plt.plot(x, y1, label = "Karmarkar Karp")
    plt.plot(x, y5, label = "Repeated Random")
    plt.plot(x, y6, label = "Hill Climbing")
    plt.plot(x, y7, label = "Simulated Annealing")
    plt.legend()
    plt.show()
    """


